File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        unordered_set<char> set1;
        int maxlen = 0;
        int l = 0, r = 0;
        while (r < n) {

            while (set1.contains(s[r])) {
                set1.erase(s[l]);
                l = l + 1;
            }
            set1.insert(s[r]);
            maxlen = max(maxlen, r - l + 1);
            
            r = r + 1;
        }
        return maxlen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        HashSet<Character> set = new HashSet<>();
        int maxlen = 0;
        int l = 0, r = 0;

        while (r < s.length()) {
            while (set.contains(s.charAt(r))) {
                set.remove(s.charAt(l));
                l = l + 1;
            }

            set.add(s.charAt(r));
            maxlen = Math.max(maxlen , r-l+1);
            r = r + 1;
        }
        return maxlen;
    }
}